src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs

```diff
@@ -190,15 +200,84 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
+                            if (metadata != null) // just to be on the safe side, should never happen
+                            {
+                                metadata.Modifications = new DynamicJsonValue(metadata) {[Constants.Documents.Metadata.ClusterTransactionIndex] = index};
+                            }
+
+                            break;
+                        case nameof(CommandType.DELETE):
+                            long deleteIndex;
+                            if (document == null ||
+                                document.TryGetWithoutThrowingOnError(Constants.Documents.Metadata.ClusterTransactionIndex,
+                                out deleteIndex) == false)
+                            {
+                                // this can happen if the user tried to delete without first loading the document
+                                // we can safely assume that such a blind write doesn't care about the state of the document
+                                // and just use the current value
+                                var cmpXngKey = CompareExchangeKey.GetStorageKey(DatabaseName, atomicGuardKey);
+                                using var _ = Slice.From(context.Allocator, cmpXngKey, out var guardKeySlice);
+                                if(items.ReadByKey(guardKeySlice, out var reader))
+                                {
+                                    deleteIndex = *(long*)reader.Read((int)ClusterStateMachine.CompareExchangeTable.Index, out var _);
+                                }
+                                else
+                                {
+                                    deleteIndex = 0;
+                                }
+                            }
+
+                            ClusterCommands.Add(new ClusterTransactionDataCommand
+                            {
+                                Type = CommandType.CompareExchangeDELETE,
+                                Id = atomicGuardKey,
+                                Index = deleteIndex,
+                                Document = context.ReadObject(new DynamicJsonValue {["Id"] = docId}, "cmp-xchg-content")
+                            });
+                            break;
+                    }
+                }
+            }
+        }
+
         public unsafe void SaveCommandsBatch(ClusterOperationContext context, long index)
         {
             if (HasDocumentsInTransaction == false)
                 return;
 
             var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
             var commandsCountPerDatabase = context.Transaction.InnerTransaction.ReadTree(ClusterStateMachine.TransactionCommandsCountPerDatabase);
-            var commands = SerializedDatabaseCommands.Clone(context);
+            var commands = context.ReadObject(SerializedDatabaseCommands, "serialized-tx-commands");
```

Comment:
Is this change needed?

Reply:
Yes, we are mutating the `SerializedDatabaseCommands` to add artifical commands here.